Core: Add Updates / Requirements for Views

nastra opened the pull request:
This PR extracts the `MetadataUpdate` / `UpdateRequirement` changes from #7913

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -23,11 +23,16 @@
 public interface MetadataUpdate extends Serializable {
   void applyTo(TableMetadata.Builder metadataBuilder);
 
+  default void applyTo(ImmutableViewMetadata.Builder viewMetadataBuilder) {}
```
Is it possible to not attach this to an Immutable interface? I'd rather have an interface that we have more control over.

rdblue replied:
I think this should throw `UnsupportedOperationException` and complain that the update cannot be applied to a view. Otherwise, incorrect use would fail silently.

rdblue replied:
@nastra, what is the benefit of Immutables here?

It looks like Immutables gives us a builder that doesn't do what we want (hide the complexity) and forces us to put logic about record keeping in the caller rather than in the builder. That's a major problem... so I'm skeptical that the right solution is to use more Immutables and have _more_ builders that add the functionality that we want.

Why don't we just scrap immutables here? It's been causing a lot of trouble.

nastra replied:
That's a fair and valid point and for builders where we'd need to have more custom logic we might actually not use generated builders and write the builder ourselves. Going forward I think there are two options:
1) leave `ViewMetadata` as Immutable but disable generating a builder. This would still perform all validation checks when a `ViewMetadata` instance is being created. This approach is currently implemented in this PR.
2) make `ViewMetadata` an actual class and implement it the same way as `TableMetadata`. This approach is implemented in https://github.com/nastra/iceberg/commit/01fd4fca5e53a371d48ba0540be449ba286eb9f2.

Let me know which one you'd prefer and I'll update this PR. For both cases we'd have our own Builder implementation.

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -60,6 +65,12 @@ public int formatVersion() {
+    @Override
+    public void applyTo(ImmutableViewMetadata.Builder viewMetadataBuilder) {
+      viewMetadataBuilder.formatVersion(formatVersion);
+      viewMetadataBuilder.addChanges(new MetadataUpdate.UpgradeFormatVersion(formatVersion));
```
Changes should be handled internally to the builder, like the `TableMetadata` builder, right?

nastra replied:
this is a bit difficult with a generated Builder unfortunately. One approach to solve this would be to define our own `Builder` that could handle adding the right metadata changes. See the implementation details of that Builder above and let me know what you think

nastra replied:
this is handled in our own builder now

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -377,5 +414,57 @@ public String location() {
+
+    @Override
+    public void applyTo(TableMetadata.Builder metadataBuilder) {
+      // noop
```
I think this should throw `UnsupportedOperationException` rather than silently ignoring the problem. If you get here, something went wrong.

nastra replied:
updated

rdblue commented on core/src/main/java/org/apache/iceberg/UpdateRequirement.java:
```diff
@@ -114,6 +117,15 @@ public void validate(TableMetadata base) {
+
+    @Override
+    public void validate(ViewMetadata base) {
+      if (base != null && base.schema().highestFieldId() != lastAssignedFieldId) {
```
This isn't correct, but it probably doesn't matter. Consider a case where I add a new field (id=3) and then roll back to a previous schema where the highest field ID is 2. That's why we store the last assigned field ID separately in the table spec. However, I don't think that we would ever use this validation for views. We don't store data files that embed schema field IDs, so we don't have any need for them to be consistent across versions or schemas. I think this should just throw an `UnsupportedOperationException ` because we should not generated the validation.

(We could also assign field IDs like we do in the table schema, but I'd probably skip that complexity)

nastra replied:
yes I agree this isn't correct, that's why I added a TODO to double-check this during review. Thanks for clarifying

rdblue commented on core/src/main/java/org/apache/iceberg/UpdateRequirement.java:
```diff
@@ -135,6 +147,15 @@ public void validate(TableMetadata base) {
+
+    @Override
+    public void validate(ViewMetadata base) {
+      if (schemaId != base.currentSchemaId()) {
```
This looks correct, but would we ever use it? You can't evolve the schema of a view because the schema must match what the version produces. That means the schema always changes with the view version (or is reused) and there isn't a need to use this validation.

rdblue commented on core/src/main/java/org/apache/iceberg/UpdateRequirement.java:
```diff
@@ -199,4 +220,56 @@ public void validate(TableMetadata base) {
+
+    @Override
+    public void validate(TableMetadata base) {
+      // noop
```
I think all the places where there are currently noops should throw.

nastra replied:
updated

rdblue commented on core/src/main/java/org/apache/iceberg/UpdateRequirement.java:
```diff
@@ -199,4 +220,56 @@ public void validate(TableMetadata base) {
     }
   }
+
+  class AssertLastAssignedViewVersionId implements UpdateRequirement {
```
What is the purpose of this validation?

For tables, we have situations where we need to assign a new ID, like `AddSchema`. In those cases, we let the service assign the ID instead of sending one so that we don't need to assert that the previous set of IDs had some maximum that we incremented. Instead, the service assigns the ID and the `SetCurrentSchema` change is sent with a `-1` to use the last assigned ID.

nastra replied:
it was mainly to express monitonically increasing view version IDs, but this can also be expressed via `AssertCurrentViewVersionID`. However, the error reporting with `AssertCurrentViewVersionID` isn't very good in indicating that view version IDs aren't monotonically increasing

rdblue replied:
Why do these need to be monotonically increasing? I don't think we care about how view version IDs are assigned, as long as there aren't duplicates.

rdblue commented on core/src/main/java/org/apache/iceberg/UpdateRequirement.java:
```diff
@@ -199,4 +220,56 @@ public void validate(TableMetadata base) {
+    }
+  }
+
+  class AssertCurrentViewVersionID implements UpdateRequirement {
```
Do we need this?

I don't think we need it because adding a new view version will assign a new ID (see the comment above) and setting the current version is an idempotent operation. We have a similar assertion, `AssertRefSnapshotId`, for snapshots, but that is because snapshots are based on one another. Here, view versions are completely independent and so replacing is an idempotent operation.

We _could_ add this if we wanted to make it possible to perform either an idempotent replace OR fail. But I don't really see a good argument for the failure case. Open to suggestion here though.

nastra replied:
> We could add this if we wanted to make it possible to perform either an idempotent replace OR fail.

Yes that was exactly the goal here and the only reason we would need this. However, I can't find a strong argument that would justify having this now. 

That being said, I think you're right and we can remove all requirements for Views for now.

rdblue commented on core/src/main/java/org/apache/iceberg/UpdateRequirement.java:
```diff
@@ -20,11 +20,14 @@
 public interface UpdateRequirement {
   void validate(TableMetadata base);
 
+  default void validate(ViewMetadata base) {}
```
Shouldn't be noop.

rdblue commented on core/src/main/java/org/apache/iceberg/UpdateRequirements.java:
```diff
@@ -117,6 +141,11 @@ private void update(MetadataUpdate.AddSchema update) {
         if (base != null) {
           require(new UpdateRequirement.AssertLastAssignedFieldId(base.lastColumnId()));
+        } else if (viewMetadataBase != null) {
+          // TODO: is this correct here? we don't track lastColumnId in ViewMetadata
```
See the comment above. Short answer is that I don't think we actually need _any_ assertions right now.

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -23,11 +23,17 @@
+import org.apache.iceberg.view.ViewVersion;
 
-/** Represents a change to table metadata. */
+/** Represents a change to table/view metadata. */
```
What about "table or view"? I'm not sure we're gaining much by using "/" instead.

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -23,11 +23,17 @@
   void applyTo(TableMetadata.Builder metadataBuilder);
 
+  default void applyTo(ViewMetadata.Builder viewMetadataBuilder) {
+    throw new UnsupportedOperationException("Applying updates to view metadata is not supported");
```
What about `"Cannot apply update %s to a view"` and use `this.getClass().getSimpleName()`? I think that's a bit more clear.

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -60,6 +66,11 @@ public int formatVersion() {
+
+    @Override
+    public void applyTo(ViewMetadata.Builder viewMetadataBuilder) {
+      viewMetadataBuilder.formatVersion(formatVersion);
```
`upgradeFormatVersion`? That's what we use for tables because you can't downgrade. It's a little unclear what we would allow right now, but simply setting the format version is probably more permissive than we need.

nastra replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -100,6 +116,11 @@ public int schemaId() {
+
+    @Override
+    public void applyTo(ViewMetadata.Builder viewMetadataBuilder) {
+      viewMetadataBuilder.currentSchemaId(schemaId);
```
`setCurrentSchemaId`?

nastra replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -343,6 +364,11 @@ public Map<String, String> updated() {
+
+    @Override
+    public void applyTo(ViewMetadata.Builder viewMetadataBuilder) {
+      viewMetadataBuilder.properties(updated);
```
`setProperties`? I think the issue with these method names is that there is no verb, so it isn't clear what is happening.

nastra replied:
right now all the builder methods are named exactly like in the generated Builder. I'll go ahead and update them to have `setXyz` so that we're in line with the Builder from `TableMetadata`

nastra replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -377,5 +408,56 @@ public String location() {
+
+    @Override
+    public void applyTo(ViewMetadata.Builder viewMetadataBuilder) {
+      viewMetadataBuilder.currentVersionId(versionId);
```
`setCurrentVersionId`

nastra replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -377,5 +408,56 @@ public String location() {
+
+    @Override
+    public void applyTo(ViewMetadata.Builder viewMetadataBuilder) {
+      viewMetadataBuilder.location(location);
```
`setLocation`?

nastra replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java:
```diff
@@ -112,6 +115,12 @@ private MetadataUpdateParser() {}
   private static final String LOCATION = "location";
 
+  // AddViewVersion
+  private static final String VIEW_VERSION = "view-version";
```
Can you update the REST spec as well?

nastra replied:
I was planning on updating the REST spec as part of https://github.com/apache/iceberg/pull/7913 because it requires a bunch of other View-specific stuff and it seems easier to update everything in one go

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+    public Builder addAllSchemas(Iterable<Schema> schemaEntries) {
+      for (Schema schema : schemaEntries) {
+        this.schemas.add(schema);
+        this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
```
Do you think we should make this optional? I don't think that `highestFieldId` is actually the last assigned field ID in all cases so it would probably be better not to send it.

nastra replied:
I think it's a good idea to make `highestFieldId` optional here. Also looks like this field is already optional in the REST Spec [here](https://github.com/apache/iceberg/blob/621d301b17417af4b71a0e8c93db9db6f8266839/open-api/rest-catalog-open-api.yaml#L1677) but not in the code. I'll open a separate PR to make this optional in the code as well, so that we don't have to set/send it for views

nastra replied:
looks like we default to `schema.highestFieldId()` when this `last-column-id` isn't send: https://github.com/apache/iceberg/blob/26af4c7387d68a04762bb50d388d3227536527b9/core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java#L404

nastra replied:
I looked into making this optional by moving `AddSchema.lastColumnId` from an  `int` to an `Integer` and not passing/sending it for views, but that is an API-breaking change unfortunately. I think we're probably ok sending `new MetadataUpdate.AddSchema(schema, schema.highestFieldId())` for Views, since that's what https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/MetadataUpdateParser.java#L404 would be doing anyway

rdblue replied:
If we aren't going to make it optional, then we need to send `max(_.highestFieldId()` for all known schemas.

nastra replied:
should be fixed now

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder schemas(Iterable<Schema> newSchemas) {
```
What does this do differently than `addAllSchemas`? Do we need a method that adds multiple schemas?

nastra replied:
the difference between `setSchemas()` and `addAllSchemas()` is that `setSchemas()` will replace whatever `this.schemas` was set to with `newSchemas`, whereas `addAllSchemas()` will just add items to `this.schemas`.

I think we're ok for now to only keep `setXyz()` and `addXyz()` without `addAllXyz`

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder currentSchemaId(Integer schemaId) {
```
Wait, do we need a current schema ID? Seems like we should always use the schema ID from the current version, so the current version ID functions as both.

nastra replied:
In that case maybe we shouldn't even send the `current-schema-id` in https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/view/ViewMetadataParser.java#L68?
That would mean that `ViewMetadata#currentSchemaId()` would not be settable and would become:
```
default Integer currentSchemaId() {
    return currentVersion().schemaId();
  }
```

nastra replied:
we discussed this offline and I will open a separate PR to remove `current-schema-id` from sending/reading it in the `ViewMetadataParser` and we won't allow setting the current schema id through the builder, because the schema id will be derived from the `ViewVersion`

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder addAllVersions(Iterable<ViewVersion> versionEntries) {
```
I don't think we need to add multiple versions at the same time.

nastra replied:
removed

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder addHistory(ViewHistoryEntry historyEntry) {
```
History should be maintained internally. We don't expect the caller to provide this.

nastra replied:
fixed now

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder properties(Map<String, String> newProperties) {
```
I think this behavior is incorrect. Setting properties should not also remove properties.

nastra replied:
fixed

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+    }
+
+    public Builder discardChanges() {
+      this.changes.clear();
```
I think this needs to be called inside the `build` method. Otherwise, the behavior changes based on when this is called.

nastra replied:
I've left this out for now until we actually need it

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+  }
+
+  class Builder {
+    private int formatVersion = 1;
```
Shouldn't this be a constant?

nastra replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -121,9 +128,178 @@ default ViewMetadata checkAndNormalize() {
+    private String location;
+    private Integer currentSchemaId;
+    private List<Schema> schemas = Lists.newArrayList();
+    private int currentVersionId = 1;
```
I think this needs to be derived from the `ViewVersion`.

nastra commented on core/src/main/java/org/apache/iceberg/view/ViewMetadataParser.java:
```diff
@@ -132,16 +133,16 @@ public static ViewMetadata fromJson(JsonNode json) {
-        .history(historyEntries)
-        .formatVersion(formatVersion)
-        .build();
+    return ImmutableViewMetadata.of(
```
switching this part so that we don't call `builder().upgradeFormatVersion()` when just setting the value

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -100,6 +117,11 @@ public int schemaId() {
+
+    @Override
+    public void applyTo(ViewMetadata.Builder viewMetadataBuilder) {
+      viewMetadataBuilder.setCurrentSchemaId(schemaId);
```
Are we going to track `current-schema-id` in views? If not, then we don't need this operation.

nastra replied:
we won't be tracking this anymore and I opened a follow-up to properly clean this up and to not have additional changes here. See also https://github.com/apache/iceberg/pull/8147#discussion_r1282023957

rdblue commented on core/src/main/java/org/apache/iceberg/MetadataUpdate.java:
```diff
@@ -377,5 +409,56 @@ public String location() {
+    @Override
+    public void applyTo(TableMetadata.Builder metadataBuilder) {
+      throw new UnsupportedOperationException(
+          String.format("Cannot apply update %s to a table", this.getClass().getSimpleName()));
```
An alternative is to make this the default implementation for `void applyTo(TableMetadata.Builder)` so that we don't need it in two places (or more if we add more view mutations).

nastra replied:
makes sense and I've done that

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+      this.changes = Lists.newArrayList(base.changes());
+    }
+
+    public Builder upgradeFormatVersion(int newFormatVersion) {
```
Shouldn't this also check that the version is <= the supported version?

Otherwise, it's possible for this builder to create an invalid change: `UpgradeFormatVersion(v)` where `v > SUPPORTED_VIEW_FORMAT_VERSION`. That may not leak right now, but it could in the future.

nastra replied:
This is being checked in https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/view/ViewMetadata.java#L87-L90, so we don't need to do it in the builder itself

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+    }
+
+    public Builder setLocation(String newLocation) {
+      if (null != this.location && this.location.equals(newLocation)) {
```
What if both are `null`? Should we use a null-safe equality check to allow setting this to `null` if it already is?

rdblue replied:
If this is required, then why would `this.location` be null?

nastra replied:
should be fixed now

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+      }
+
+      this.location = newLocation;
+      this.changes.add(new MetadataUpdate.SetLocation(newLocation));
```
Related to the change above, can this be null?

nastra replied:
technically you could pass `null`, but we won't be able to create a `ViewMetadata` instance, because the API requires `location` to be non-null

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder setCurrentVersionId(int newVersionId) {
```
I think this should check that the version exists.

nastra replied:
this check is already done in https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/view/ViewMetadata.java#L96-L100

rdblue replied:
@nastra, I think this needs to be done here for the reasons we talked about.

> it should not affect and should not cause any issues when re-applying SetCurrentViewVersion, because that then translates to the same Builder calls that were done to cause the changes in the first place

This is a mistaken assumption. There is no guarantee that a REST catalog service is using this code. We cannot assume things like that and should always produce requests that are reasonable and correct.

> this would require to always set the list of versions first on the Builder before one would be able to call setCurrentVersionId(), which would result in awkward API behavior.

Many other builders work this way so I don't think it is too awkward. When there are no dependencies between calls then allowing an arbitrary order is definitely a best practice. But when there are dependencies it's reasonable to expect that a version has been added before it is referenced.

rdblue replied:
I think we also need to validate that the schema ID of the new version is present in this method.

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder setCurrentSchemaId(Integer newSchemaId) {
```
I don't think we need this.

nastra replied:
agreed, and it will be removed as part of https://github.com/apache/iceberg/pull/8210 once this PR is in

rdblue replied:
Let's remove it now rather than adding methods that have no purpose and will be removed.

nastra replied:
fair enough, I've removed it as part of https://github.com/apache/iceberg/pull/8210 and rebased this PR, meaning that this PR won't have any `current-schema-id` changes anymore

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+    }
+
+    public Builder addVersion(ViewVersion version) {
+      this.versions.add(version);
```
This should also update the index of versions by ID so that validation for setting the current version ID works.

nastra replied:
this is being handled by `checkAndNormalize()`, so I don't think we need anything in the builder itself for this

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder setVersions(Iterable<ViewVersion> versionsToAdd) {
```
Looks like this isn't named correctly. It isn't clear what "set" would mean. Luckily, I don't think we need a method to add multiple versions at once, do we?

nastra replied:
previously we were mainly using it in `ViewMetadataParser`, but that's not the case anymore, so I've removed it for now.

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder addProperty(String key, String value) {
```
The verb "add" is misleading here because there is no guarantee that it doesn't already exist. Using "set" is clear because it is idempotent.

nastra replied:
I've removed this method for now until we actually need it

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,187 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder setProperties(Map<String, String> propertiesToAdd) {
```
How about `newProperties` instead? Using the verb "add" is misleading here as well.

nastra replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadataParser.java:
```diff
@@ -132,16 +133,16 @@ public static ViewMetadata fromJson(JsonNode json) {
+        versions,
+        historyEntries,
+        properties,
+        ImmutableList.of());
```
What about the view metadata location? We track that in `TableMetadata` so we can determine if it changed.

nastra replied:
we haven't been tracking that for Views so far but it was on my list to be added once we need it. That being said, I'll handle this in a separate PR

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -117,13 +128,172 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public ViewMetadata build() {
```
Shouldn't this also enforce the number of versions? That way we don't need the validation that also modifies the object that is constructed?

nastra replied:
the problem with enforcing the number of versions in the `build()` method is that we'd have to enforce this in some other place as well, because `ViewMetadataParser` doesn't go through the builder, but we want to have the same behavior, independent from whether an instance is created through the builder or directly via `ImmutableViewMetadata.of(...)`

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+      return this;
+    }
+
+    public Builder setSchemas(Iterable<Schema> schemasToAdd) {
```
This has the same problem that I pointed out about `setVersions` here: https://github.com/apache/iceberg/pull/8147#pullrequestreview-1575884865. There's no need to have this method, so let's remove it.

nastra replied:
done

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+      return builder.build();
+    }
+
+    private static Map<Integer, Schema> indexSchemas(List<Schema> schemasToIndex) {
```
Why are these static methods inside the builder?

nastra replied:
we're using them in `ViewMetadata` + in the `Builder` and making them static is the only way to share the logic for both

rdblue replied:
Sure, but why are they inside the builder? What is the purpose of scoping these methods to the builder and then using them outside?

rdblue replied:
Ah, I see. Because this is using Immutables, the outer scope is an interface where helper methods can't be made private.

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+
+    public Builder addSchema(Schema schema) {
+      this.schemas.add(schema);
+      this.changes.add(new MetadataUpdate.AddSchema(schema, schema.highestFieldId()));
```
This isn't correct. This needs to send the max of the highest field ID across all known schemas, not just the new schema.

nastra replied:
I've moved building up the changes into `build()` where we can iterate over all schemas and figure out the highest field ID

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+
+    public ViewMetadata build() {
+      Preconditions.checkArgument(null != location, "Invalid location: null");
+      Preconditions.checkArgument(versions.size() > 0, "Invalid view versions: empty");
```
While this is technically true, it doesn't really point to the problem, which is that `addVersion` was never called. I think this should be something like "Invalid view: no versions were added".

nastra replied:
adjusted the error msg for `versions` and `schemas`

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+      Preconditions.checkArgument(null != location, "Invalid location: null");
+      Preconditions.checkArgument(versions.size() > 0, "Invalid view versions: empty");
+      Preconditions.checkArgument(history.size() > 0, "Invalid view history: empty");
+      Preconditions.checkArgument(schemas.size() > 0, "Invalid schemas: empty");
```
We probably don't need this. The current version's schema should be checked in this method, which will ensure that at least one was added.

nastra replied:
for consistency I think having `Invalid view: no schemas were added` makes sense, as it immediately points out what the problem is (no schemas were added), while `Cannot find current schema with id 1 in schemas: []` is less explicit about that.

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+    public ViewMetadata build() {
+      Preconditions.checkArgument(null != location, "Invalid location: null");
+      Preconditions.checkArgument(versions.size() > 0, "Invalid view versions: empty");
+      Preconditions.checkArgument(history.size() > 0, "Invalid view history: empty");
```
This is tracked internally, so if there is a problem, it's a bug. No need to throw an exception that the caller can't do anything about.

nastra replied:
makes sense, removed it

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+              ViewProperties.VERSION_HISTORY_SIZE,
+              ViewProperties.VERSION_HISTORY_SIZE_DEFAULT);
+
+      if (versionHistorySizeToKeep <= 0) {
```
Now that this is in the builder, we can throw an exception if it is invalid. Maybe convert this to a precondition?

nastra replied:
makes sense, added a precondition and a test to ensure this fails properly

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+                .collect(Collectors.toSet());
+        changesToKeep.removeAll(toRemove);
+
+        return ImmutableViewMetadata.of(
```
We usually prefer having fewer return statements and using a variable like `outputVersions` or `finalHistory` that gets updated depending on the path.

nastra replied:
done

nastra commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +109,215 @@ default ViewMetadata checkAndNormalize() {
+      }
+
+      this.currentVersionId = newVersionId;
+      this.changes.add(new MetadataUpdate.SetCurrentViewVersion(newVersionId));
```
I believe this should be mostly addressed in the latest few changes

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
-    Preconditions.checkArgument(versions().size() > 0, "Invalid view versions: empty");
-    Preconditions.checkArgument(history().size() > 0, "Invalid view history: empty");
-    Preconditions.checkArgument(schemas().size() > 0, "Invalid schemas: empty");
+    return this;
```
Why does this method return `this`? Other check methods are defined as `default void check() { ... }`

nastra replied:
this was because we initially did history **normalization** in this method, which required to return either `this` or a new instance

rdblue replied:
Thanks for catching this! I missed it in my PR.

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+    private List<Schema> schemas = Lists.newArrayList();
+    private int currentVersionId;
+    private Integer lastAddedVersionId;
+    private Integer lastAddedSchemaId;
```
If these are intended to default to null, it is better to explicitly add that default here.

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+      // changes can be added
+      int highestFieldId = highestFieldId();
+      for (Schema schema : schemas) {
+        this.changes.add(new MetadataUpdate.AddSchema(schema, highestFieldId));
```
Why would this add all of the known schemas? I don't think this should be done here.

nastra replied:
should be fixed now with your changes

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+    }
+
+    public Builder setCurrentVersion(ViewVersion version) {
+      Schema schema = indexSchemas(schemas).get(version.schemaId());
```
Why does this index schemas every time a new version is added?

nastra replied:
should be fixed now with your changes

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+
+    public Builder setCurrentVersion(ViewVersion version) {
+      Schema schema = indexSchemas(schemas).get(version.schemaId());
+      int newSchemaId = addSchemaInternal(schema);
```
This doesn't seem correct to me.

The `ViewVersion` references a schema by ID. This method assumes that schema has been added to the builder already (or was already added to the view in a previous commit) because it expects the schema to be present in the ID to schema map produced by `indexSchemas`. But this then adds the schema and possibly reassigns the schema ID, which will be a noop because the schema is already known.

I think the solution is to pass the schema into this method. Then this would add the schema via `addSchemaInternal` (possibly reassigning its ID), add the view version that references the schema ID returned by `addSchemaInternal`, and then call `setCurrentVersionId` with the view version ID returned by `addVersionInternal`.

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+        boolean isNewVersion =
+            lastAddedVersionId != null
+                && changes(MetadataUpdate.AddViewVersion.class)
+                    .anyMatch(added -> added.viewVersion().versionId() == newVersionId);
```
The formatting here makes this really hard to read. I think it is better in cases like this to separate expressions and use more variables. Even if you have to split the expression across lines, at least it isn't mixing boolean expression formatting with formatting for long call chains.

```java
boolean isAddedVersion = changes(MetadataUpdate.AddViewVersion.class)
    .anyMatch(added -> added.viewVersion().versionId() == newVersionId);
boolean isNewVersion = lastAddedVersionId != null && isAddedVersion;
```

That said, I don't understand why it works this way. If `addVersionInternal` is called with a view version that already exists, then it should return the existing ID and not set `lastAddedVersionId` because no version was added. That's a simpler way to do what I think you're doing here.

nastra replied:
should be fixed now with your changes

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+        newVersion = version;
+      }
+
+      this.versions.add(newVersion);
```
When referencing an instance field, there is no need to use a `this.` prefix. Only assignment should.

nastra replied:
should be fixed now with your changes

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+            lastAddedVersionId != null
+                && changes(MetadataUpdate.AddViewVersion.class)
+                    .anyMatch(added -> added.viewVersion().versionId() == newVersionId);
+        this.lastAddedVersionId = isNewVersion ? newVersionId : null;
```
I think this is trying to account for a case that should never happen, where `addViewVersion` is called with an existing version and then `setCurrentVersionId` is called with `-1`. But `-1` should only be sent to set the last view version if that version's ID is unknown -- that is, when it is a new view version and not an existing view version. It should be sufficient to match the logic in `TableMetadata`.

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+    }
+
+    private int addSchemaInternal(Schema schema) {
+      if (schema.schemaId() == -1) {
```
I don't think this is handling a valid case. While the REST protocol uses -1 to signal the last added schema, it would be meaningless for a schema to use that ID.

nastra replied:
I agree that this wasn't really handling a valid case. This was rather an open TODO that we discussed internally. should be fixed now with your changes

rdblue commented on core/src/main/java/org/apache/iceberg/view/ViewMetadata.java:
```diff
@@ -91,41 +111,325 @@ default ViewMetadata checkAndNormalize() {
+            lastAddedSchemaId != null, "Cannot set last added schema: no schema has been added");
+
+        return addSchemaInternal(
+            new Schema(lastAddedSchemaId, schema.columns(), schema.identifierFieldIds()));
```
If there is no last added schema ID -- which is set by adding a schema -- then this will fail with a `NullPointerException` that has no context. Because this is set by adding a schema, the only time this would succeed is when you've added a schema with a "normal" ID (e.g. 5) and then add a schema with ID -1 for some reason.

It gets weirder from there because this would use the last added schema ID to add a new schema with possibly different fields. Luckily, the schema to add would be sent through `reuseOrCreateNewSchemaId`.

As I noted above, I don't think it is valid to have a schema with `schemaId=-1` and to handle it differently. The correct behavior is to add the schema and control schema ID assignment in this builder, which is already done by `reuseOrCreateNewSchemaId`.

nastra replied:
should be fixed now with your changes

rdblue commented:
@nastra, to avoid more review rounds, I went ahead and fixed the remaining issues in this PR as I was doing a deep dive. See https://github.com/nastra/iceberg/pull/135 with the changes.

nastra commented:
> @nastra, to avoid more review rounds, I went ahead and fixed the remaining issues in this PR as I was doing a deep dive. See [nastra#135](https://github.com/nastra/iceberg/pull/135) with the changes. You may need to fix a test or two, but I got the `ViewMetadata` tests passing.

Thanks @rdblue, I've updated a few small things as well and pushed them.

rdblue commented:
Merged. Thanks, @nastra!
